Subscribe push notification

NikolaySlavov-96 opened the pull request.

NikolaySlavov-96 committed:
feat: Implement PushNotification entity and enhance user-related entities

NikolaySlavov-96 committed:
feat: Implement PushNotification repository and interface

NikolaySlavov-96 committed:
feat: Enhance UserRepository with GetIdByDeviceId method

NikolaySlavov-96 committed:
feat: Implement PushNotification service and DTOs

NikolaySlavov-96 committed:
feat: Implement GET, POST, and PUT requests for PushNotification

NikolaySlavov-96 committed:
feat: Implement CustomClaimTypes in extension

NikolaySlavov-96 committed:
feat: Include custom claims and DeviceId in access and refresh tokens

NikolaySlavov-96 committed:
refactor: Update logic, rename services and interfaces, and add new interfaces

NikolaySlavov-96 committed:
refactor: Enhance POST subscribe request logic

- Improved logic to extract DeviceId directly from the token, streamlining the subscription process.
- Removed the returned response for the ID upon successful subscription creation to simplify the API response.

NikolaySlavov-96 committed:
chore: Remove unused row from entity

NikolaySlavov-96 committed:
feat: Extract operating system name into separate table and update PushNotificationTable

- Extracted operating system names into a dedicated table to normalize data and improve database design.
- Added a migration to implement the new table structure and update related foreign key relationships.
- Updated PushNotificationTable to reference the new operating system table, ensuring data consistency.

NikolaySlavov-96 committed:
feat: Implement seeder mechanism for initializing static data in new databases

- Developed a seeder mechanism to automatically populate new databases with essential static data.
- Implemented a specific seed for the OperationSystem entity to ensure all necessary operating system entries are present.
- Enhanced database initialization by ensuring consistent and complete data setup for new deployments.

NaumVitanov commented on Database/Configurations/OperationSystemEntityConfiguration.cs:
```diff
@@ -0,0 +1,14 @@
+    {
+        public void Configure(EntityTypeBuilder<OperationSystemEntity> builder)
+        {
+            builder.ToTable("OperationSystem");
```
The table name should be plural.

NaumVitanov commented on Database/Entities/OperationSystemEntity.cs:
```diff
@@ -0,0 +1,16 @@
+
+        [Required]
+        public string TypeOS { get; set; }
+
```
We should not be leaving warnings like this one. It opens up the possibility for bugs.

NaumVitanov commented on Database/Context/AppDbContext.cs:
```diff
@@ -16,6 +16,8 @@ public class AppDbContext(DbContextOptions<AppDbContext> options) : DbContext(op
 	public DbSet<MessageEntity> Messages { get; set; }
 	public DbSet<MessageStatusEntity> MessagesStatus { get; set; }
+	public DbSet<PushNotificationEntity> PushNotification { get; set; }
+	public DbSet<OperationSystemEntity> OperationSystem { get; set; }
```
Both table names should be in plural, here and in the config of each entity.

NaumVitanov commented on Database/Entities/PushNotificationEntity.cs:
```diff
@@ -0,0 +1,43 @@
+        [Required]
+        public string Token { get; set; }
+
+        public bool IsTurnOnNotification { get; set; }
```
"Enabled" would be better for the naming of that property. If you do not find it descriptive enough, you can make it "NotificationEnabled" or even better write a short summary above it to explain what it does.

NaumVitanov commented on Database/Entities/PushNotificationEntity.cs:
```diff
@@ -0,0 +1,43 @@
+        public bool IsTurnOnNotification { get; set; }
+
+        [ForeignKey(nameof(UserId))]
+        public virtual UserEntity? User { get; set; }
```
Why is the user nullable? You are setting the userId from the constructor is it possible for this property to be null?

NaumVitanov commented on Database/Seeders/Seeds/OperationSystemSeeder.cs:
```diff
@@ -0,0 +1,25 @@
+        {
+            if (!context.OperationSystem.Any())
+            {
+                var entity = new List<OperationSystemEntity>
```
Use explicit types and then you can make it "= new()". That's the approach used in the project.

NaumVitanov commented on Database/Configurations/PushNotificationEntityConfiguration.cs:
```diff
@@ -0,0 +1,35 @@
+                .HasOne(e => e.User)
+                .WithMany(e => e.PushNotification)
+                .HasForeignKey(e => e.UserId)
+                .HasPrincipalKey(e => e.Id)
```
This method is written ONLY for clarity in this case. When using the primary key as a connection there is no need to call that method.

NikolaySlavov-96 replied:
We will discuss it in the conversation.

NaumVitanov commented on Database/Entities/PushNotificationEntity.cs:
```diff
@@ -0,0 +1,43 @@
+        public int DeviceId { get; set; }
+
+        [Required]
+        public int OS { get; set; }
```
This should be named OsId. Please follow the convention.

NikolaySlavov-96 replied:
In this case, if I name it "Os," would you understand in two months that it refers to the commonly accepted name for an operating system, or would you wonder what is meant?

vgechev replied:
I'm not sure why you are shortening it at all. You can call it `OperatingSystemId`

NaumVitanov replied:
I agree with Valentino here. On the other hand, even if we leave it at Os, we can write a summary of the property which will then be available every time someone puts the cursor on the property anywhere in the code. Both ways it will be understandable.

NikolaySlavov-96 replied:
OperatingSystemType OperatingSystem. 💯

NaumVitanov commented on Database/Entities/PushNotificationEntity.cs:
```diff
@@ -0,0 +1,43 @@
+{
+    public class PushNotificationEntity
+    {
+        public PushNotificationEntity(int oS, string token, bool IsTurnOnNotification, int UserId, int deviceId)
```
This naming suggest that each row in the table will hold one push notification. If I understand the logic correctly, this entity and table hold the configurations of a push notification settings for a given device. Please name the entity accordingly, or correct me if I am wrong about the logic here.

NikolaySlavov-96 replied:
How would you like me to name it?

NaumVitanov replied:
I purposefully did not suggest a name. My idea was for you think of an appropriate one. In the sake of answering your question, here is a suggestion for the naming, which is not perfect but a bit more accurate: DeviceNotificationConfigEntity

NikolaySlavov-96 replied:
How do you feel about remaining all of this in the other PR that is for integrating the sending of notifications?

NaumVitanov commented on Database/Configurations/PushNotificationEntityConfiguration.cs:
```diff
@@ -0,0 +1,35 @@
+                .HasOne(e => e.UserDevice)
+                .WithOne(e => e.PushNotification)
+                .HasForeignKey<PushNotificationEntity>(e => e.DeviceId)
+                .HasPrincipalKey<UserDeviceEntity>(e => e.Id)
```
This looks strange to me. Why do we have to specify the entity for the FK and the PK in one to one relationship, but with one to many we do not have to specify it? Are we missing something or is there no other way to do that?

NikolaySlavov-96 replied:
The conversation in "Telegram" about which approach to keep and which will continue in the project.

NaumVitanov commented on Database/Entities/UserDeviceEntity.cs:
```diff
@@ -36,4 +36,5 @@ public UserDeviceEntity(UserEntity user, string deviceId)
 	[ForeignKey(nameof(UserId))]
 	public virtual UserEntity? User { get; set; }
 	public virtual UserLoginCodeEntity? UserLoginCode { get; set; }
+	public virtual PushNotificationEntity? PushNotification { get; set; }
```
Why is there no config for this virtual property? How is the connection made?

NikolaySlavov-96 replied:
Can you give me guidance on how to configure it?

NaumVitanov replied:
Yes, let's do that over a call, if it's ok with you.

NaumVitanov commented on Services/Extensions/ClaimsPrincipalExtensions.cs:
```diff
@@ -11,4 +11,15 @@ public static int Id(this ClaimsPrincipal user)
+	public static string DeviceId(this ClaimsPrincipal user)
+	{
+		string? deviceId = user.FindFirstValue(CustomClaimTypes.DeviceId);
+		return deviceId ?? throw new ArgumentNullException("Could not extract user Device ID from token");
```
Please fix the message to be in the same style as the other one for the user id. It would look something like this:

"Could not extract device id from token"

NaumVitanov commented on Services/Extensions/ClaimsPrincipalExtensions.cs:
```diff
@@ -11,4 +11,15 @@ public static int Id(this ClaimsPrincipal user)
+
+public static class CustomClaimTypes
+{
+	public static readonly string DeviceId = "http://schemas.yourapp.com/claims/deviceid";
```
Can you make this as an extension to the the ClaimTypes class?

NaumVitanov commented on Services/Repositories/OperationSystem/Repository/OperationSystemRepository.cs:
```diff
@@ -0,0 +1,25 @@
+    {
+        public async Task<int?> GetOperationSystemId(string operationSystem)
+        {
+            OperationSystemEntity result = await dbContext.OperationSystem.FirstOrDefaultAsync(o => o.TypeOS == operationSystem);
```
This should be nullable.

NikolaySlavov-96 replied:
What do you mean?

NaumVitanov replied:
Since the project uses nullable annotations and warnings, this object of type OperationSystemEntity should be nullable. FirstOrDefault, could return null, hence the nullable type. It would look like that:  OperationSystemEntity? result = await dbContext.OperationSystem.FirstOrDefaultAsync. Please notice the question mark added after the class.

NaumVitanov commented on Services/Repositories/OperationSystem/Repository/OperationSystemRepository.cs:
```diff
@@ -0,0 +1,25 @@
+
+        public Task SaveChangesAsync()
+        {
+            throw new NotImplementedException();
```
That should be implemented. It should actually be implemented in a base repo which should be inherited by this repo, because it's needed in every repo abstraction possible

NikolaySlavov-96 replied:
Please verify!

NaumVitanov replied:
Could you please clarify what type of verification is needed here? We could do it over a call, for easier communication.

NaumVitanov commented on Services/Repositories/PushNotification/Repository/PushNotificationRepository.cs:
```diff
@@ -0,0 +1,45 @@
+        {
+            notificationEntity = dbContext.PushNotification.Add(notificationEntity).Entity;
+
+            await SaveChangesAsync();
```
How are you handling the logic if this row throws when saving?

NikolaySlavov-96 replied:
Based on the code that exists in this project, I don't see anywhere where such a type of error is being caught. If you mean a "try-catch" block, I don't think that would be necessary because an error will be returned to the service, and it will, in turn, return it to the controller, which will return something other than "success" to the user.

NaumVitanov replied:
I do not say that because of how anywhere else in the project this type of case is handled.  At the time of writing this comment it seemed to me, that if we leave it like that, a possibility for a problem is left open. When looking at the code again, I see that you are right to say that the normal flow will handle it good enough.

NaumVitanov commented on Services/PushNotification/Models/ChangeStatusRequest.cs:
```diff
@@ -0,0 +1,3 @@
+namespace Services.PushNotification.Models;
+
+public record ChangeStatusRequest(bool IsNotificationStatus);
```
Change this property with the same name as the one I suggested for the DB property. -  "Enabled"

NaumVitanov commented on Services/PushNotification/Models/PushNotificationResponseDeviceData.cs:
```diff
@@ -0,0 +1,3 @@
+namespace Services.PushNotification.Models;
+
+public record PushNotificationResponseDeviceData(bool IsTurnOnNotification);
```
Same naming as the two before

NaumVitanov commented on Services/Repositories/PushNotification/Repository/PushNotificationRepository.cs:
```diff
@@ -0,0 +1,45 @@
+        }
+
+        public Task<bool> DoesDeviceTokenExist(PushNotificationDTO notificationDTO, int userId) =>
+            dbContext.PushNotification.AnyAsync(p => p.Token == notificationDTO.Token && p.UserId == userId);
```
The naming of the method is wrong. It checks if there is a token for the given user, not if the token exists for any user.

NikolaySlavov-96 replied:
Done! Please verify

NaumVitanov commented on Services/Repositories/User/Repository/UserRepository.cs:
```diff
@@ -77,4 +77,7 @@ public Task<bool> DoesUserExist(int id) =>
 	public Task<UserEntity[]> GetUsers(int[] userIds) =>
 		dbContext.Users.Where(u => userIds.Contains(u.Id)).ToArrayAsync();
+
+	public Task<UserDeviceEntity?> GetIdByDeviceId(string deviceId) =>
```
Wrong naming. You are getting the device by the deviceId, not Id by the deviceId as you have written

NikolaySlavov-96 replied:
Oooo. Thanks!

NaumVitanov commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+            var userDevice = await userRepo.GetIdByDeviceId(deviceId);
+            if (userDevice == null)
+            {
+                return Result.Failure("Device Id doesn't exists");
```
I do not think the FE (the user) needs so much details in the errors. Leave fewer errors which combine the different scenarios.

NikolaySlavov-96 replied:
Yes, the user does not need this message, but I think it is necessary for easier understanding of the problem in special cases. In the case when you have only one general message for a problem with only one specific user ID, it would be much harder to find the problem on their phone compared to if we have a clear and precise message about where things went wrong.

NaumVitanov replied:
Aren't these messages used only to be send to the FE or am I missing something here? Those are not the error logs...

NikolaySlavov-96 replied:
I was under the impression that if something like "Sentry" is used in the future, we could more easy understand  what the problem was between the backend and frontend.

NaumVitanov commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+                return Result.Failure($"This operating system {deviceData.OS} is not maintained");
+            }
+
+            PushNotificationEntity notificationEntity = new(operationSystemId.Value, deviceData.Token, IsTurnOnNotification: true, UserId: userId, userDevice.Id);
```
You have a relation between the devices and the notifications, also you have a relation between the OS and the notification (using the OS itself). You can get all of that info in one request to the DB instead of three separate ones.

NikolaySlavov-96 replied:
I'm sorry, I don't understand you. In this case, before creating a record, I check if the data exists, and if it does, I return a descriptive and precise error, and I also ensure that data duplication does not occur. If you mean to just make one query and add several "&&", for me, it becomes too complicated. I like simple things.

NaumVitanov replied:
Simple is good! Maybe my idea is overcomplicating things here. Good one :)

NikolaySlavov-96 replied:
Please check!

NaumVitanov reviewed (commented):
Good job! Could use some improvements. Details about the PR comments: Some of the comments I will be writing only once, but they will be applicable to more than one place. Please check your whole logic according to all of my comments. E.g. when I type a comment about using explicit types, it applies to the whole project not just this line. Another example is the naming of a table. If I write it should be plural that applies to everywhere it is used not only where my comment is.

NikolaySlavov-96 committed:
comm: Rename methods and variables, update return messages

NikolaySlavov-96 committed:
refactor: Rename column and tables, add migration for database updates

- Renamed column in PushNotifications table from IsTurnOnNotification to IsNotificationEnabled for clarity and consistency.
- Changed table names from PushNotification and OperationSystem to PushNotifications and OperationSystems to follow naming conventions.
- Added a migration script to apply these changes to the database schema, ensuring a smooth transition.
- Updated all relevant code references to reflect the new column and table names.

vgechev commented on Database/Configurations/OperationSystemEntityConfiguration.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Database.Configurations
+{
+    internal class OperationSystemEntityConfiguration : IEntityTypeConfiguration<OperationSystemEntity>
```
"Entity" as part of the name is redundant. It's only clutter, same goes for Configuration. You can easily see it's a config from the namespace.

vgechev commented on Database/Configurations/OperationSystemEntityConfiguration.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    internal class OperationSystemEntityConfiguration : IEntityTypeConfiguration<OperationSystemEntity>
+    {
+        public void Configure(EntityTypeBuilder<OperationSystemEntity> builder)
```
Entity in the name is redundant here as well.

vgechev commented on Database/Configurations/PushNotificationEntityConfiguration.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace Database.Configurations
+{
+    internal class PushNotificationEntityConfiguration : IEntityTypeConfiguration<PushNotificationEntity>
```
same here (naming).

vgechev commented on Database/Configurations/PushNotificationEntityConfiguration.cs:
```diff
@@ -0,0 +1,35 @@
+{
+    internal class PushNotificationEntityConfiguration : IEntityTypeConfiguration<PushNotificationEntity>
+    {
+        public void Configure(EntityTypeBuilder<PushNotificationEntity> builder)
```
same here (naming).

vgechev commented on Database/Entities/OperationSystemEntity.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public class OperationSystemEntity
+    {
+        public OperationSystemEntity(string typeOS)
```
possible use of expression

vgechev commented on Database/Entities/OperationSystemEntity.cs:
```diff
@@ -0,0 +1,21 @@
+        [Key]
+        public int Id { get; set; }
+
+        [Required]
```
Let's be consistent and use a single way of defining required properties. I can see in other cases we are using the configuration classes to define these:

![image](https://github.com/user-attachments/assets/7d3c6295-6604-45a1-a7c1-0c4f150b4480)


Same goes for [Key]

NikolaySlavov-96 replied:
In this project, given its small size, i have noticed at least three approaches to working with the database. Util there is a clarified way of working that will be used. consider this comment as skipped.

vgechev commented on Database/Entities/PushNotificationEntity.cs:
```diff
@@ -0,0 +1,43 @@
+{
+    public class PushNotificationEntity
+    {
+        public PushNotificationEntity(int oS, string token, bool IsNotificationEnabled, int UserId, int deviceId)
```
os is better than oS, IMO, and operatingSystem is the best.

vgechev commented on Database/Entities/OperationSystemEntity.cs:
```diff
@@ -0,0 +1,21 @@
+    {
+        public OperationSystemEntity(string typeOS)
+        {
+            TypeOS = typeOS;
```
OSType, or even better - OperatinglSystemType

vgechev commented on Database/Entities/PushNotificationEntity.cs:
```diff
@@ -0,0 +1,43 @@
+        [Required]
+        public int OS { get; set; }
+
+        [Required]
```
. (refer to prev comment)

vgechev commented on Database/Migrations/20241231094258_addPushNotificationWithSeparateTableFroNames.cs:
```diff
@@ -0,0 +1,89 @@
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.CreateTable(
+                name: "OperationSystem",
```
It's usually called "Operating system" not *operation* system.

vgechev commented on Services/Extensions/ClaimsPrincipalExtensions.cs:
```diff
@@ -11,4 +11,15 @@ public static int Id(this ClaimsPrincipal user)
+
+public static class CustomClaimTypes
+{
+	public static readonly string DeviceId = "http://schemas.yourapp.com/claims/deviceid";
```
Is this really the best name for this claim?

NikolaySlavov-96 replied:
What would be a better name for this?

vgechev commented on Services/Extensions/ServiceCollectionExtension.cs:
```diff
@@ -160,6 +169,11 @@ public static IServiceCollection AddServices(this IServiceCollection services) =
+			.AddScoped<IPushNotification, PushNotificationService>()
+			.AddScoped<IPushNotificationRepository, PushNotificationRepository>()
+			.AddScoped<IOperationSystemRepository, OperationSystemRepository>()
+			.AddTransient<ISeeder, OperationSystemSeeder>()
```
.

NikolaySlavov-96 replied:
What should "." convey?

vgechev commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+{
+    public class PushNotificationService(IPushNotificationRepository pushNotificationRepo, IUserRepository userRepo, IOperationSystemRepository operationSystemRepo) : IPushNotification
+    {
+        public async Task<Result> SubscribeForPushNotification(PushNotificationDTO deviceData, int userId, string deviceId)
```
"deviceData" i love how you simplified and shortened the name here, but in the interface it's different and longer. This one is better, or you can even use only "data".

NikolaySlavov-96 replied:
I don't understand which interface you are talking about that has a different name

vgechev commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+    {
+        public async Task<Result> SubscribeForPushNotification(PushNotificationDTO deviceData, int userId, string deviceId)
+        {
+            bool checkDeviceTokenExists = await pushNotificationRepo.CheckDeviceTokenExists(deviceData, userId);
```
I prefer "Does Device Token Exist"

NikolaySlavov-96 replied:
This is the name I used the first time @NaumVitanov

vgechev commented on Services/Repositories/PushNotification/Repository/PushNotificationRepository.cs:
```diff
@@ -0,0 +1,45 @@
+            return notificationData;
+        }
+
+        public Task<bool> CheckDeviceTokenExists(PushNotificationDTO notificationDTO, int userId) =>
```
I prefer "Does Device Token Exist". Also, why are you passing the whole dto when you are expecting only the token, it makes no sense.

vgechev commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+            bool checkDeviceTokenExists = await pushNotificationRepo.CheckDeviceTokenExists(deviceData, userId);
+            if (checkDeviceTokenExists)
+            {
+                return Result.Failure("This Device token already exists");
```
I don't understand the logic here. Why do we have device tokens? Why are we checking for them, etc..?
Please, let's start writing summaries and comments and explanations in the code. This way everyone will be on the same page and we'll have fewer questions.

NikolaySlavov-96 replied:
I understand that you want to describe how the entire process works, what Apple and Android need to know in order to understand whom to send a notification. Why do we actually need to check if the same user is trying, for some reason, to subscribe with the same device twice?

NikolaySlavov-96 replied:
In "PushNotificationDTO" i'm adding description of what the two properties mean.

vgechev commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+            var userDevice = await userRepo.GetDeviceByDeviceId(deviceId);
+            if (userDevice == null)
+            {
+                return Result.Failure("Device Id doesn't exists");
```
Shouldn't we check first if the device exists and then for a token for it?

vgechev commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+{
+    public class PushNotificationService(IPushNotificationRepository pushNotificationRepo, IUserRepository userRepo, IOperationSystemRepository operationSystemRepo) : IPushNotification
+    {
+        public async Task<Result> SubscribeForPushNotification(PushNotificationDTO deviceData, int userId, string deviceId)
```
You are using a DTO with only two properties, why? Either include all the parameters in it and just pass the DTO or remove the DTO and pass 4 params. I don't understand the point of splitting it like so.

NikolaySlavov-96 replied:
I thought that what comes and is expected from the body of the request is more appropriate to be separated from the things that are taken from the "accessToken" itself and passed internally.

NikolaySlavov-96 replied:
At this moment i think that this way it would be readable, and i will keep it this way.

vgechev commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+            var operationSystemId = await operationSystemRepo.GetOperationSystemId(deviceData.OS);
+            if (operationSystemId == null)
+            {
+                return Result.Failure($"This operating system {deviceData.OS} is not maintained");
```
is not supported*

vgechev commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+            if (operationSystemId == null)
+            {
+                return Result.Failure($"This operating system {deviceData.OS} is not maintained");
+            }
```
When you have multiple checks (more than 1, imo), I think it's way better to just extract them in a method, call it ValidateData or whatever, which returns an array of errors or the first validation error it hits, and then in the calling method to check if there are any errors returned from the ValidateData method and if so, return Result.Failure(errors). It will improve the readability by a lot.

NikolaySlavov-96 replied:
At this moment i think that this way it would be readable, and i will keep it this way.

vgechev commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+                return Result.Failure($"This operating system {deviceData.OS} is not maintained");
+            }
+
+            PushNotificationEntity notificationEntity = new(operationSystemId.Value, deviceData.Token, IsNotificationEnabled: true, UserId: userId, userDevice.Id);
```
no need to have "UserId: ", just pass the userId

vgechev commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+            PushNotificationEntity? result = await pushNotificationRepo.UpdateDeviceData(userId, deviceId, request.IsNotificationEnabled);
+            if (result == null)
+            {
+                return Result.Failure("No data found or new record is the same with old record");
```
.. is the same **as**  old ..

vgechev commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+
+        public async Task<Result> ChangePushNotificationStatus(int userId, string deviceId, ChangeStatusRequest request)
+        {
+            PushNotificationEntity? result = await pushNotificationRepo.UpdateDeviceData(userId, deviceId, request.IsNotificationEnabled);
```
Isn't it better if this method returns a boolean instead? Just true if successful, false if not. Why do you need the whole entity (tracked)?

NikolaySlavov-96 replied:
After the changes made, we no longer need it. Thank you very much for your attentiveness.

vgechev commented on Services/PushNotification/Service/PushNotificationService.cs:
```diff
@@ -0,0 +1,63 @@
+            return Result.Success();
+        }
+
+        public async Task<Result<PushNotificationResponseDeviceData>> GetPushNotificationData(string deviceId, int userId)
```
What is happening here? Why are you using and passing the whole entity again when you only need a boolean in the end, to check whether or not the notification is enabled for that user and device? Don't use entities like this, for nothing, they're very expensive in terms of resources, especially when they are tracked by the db context.

vgechev replied:
dbContext.PushNotifc.Where(x => x...).Select(x => x...).SingleOrDefaultAsync();

vgechev commented on Services/Repositories/PushNotification/Repository/PushNotificationRepository.cs:
```diff
@@ -0,0 +1,45 @@
+            notificationEntity = dbContext.PushNotifications.Add(notificationEntity).Entity;
+
+            await SaveChangesAsync();
+            return notificationEntity;
```
Why are you returning the entity here? You are not using it anywhere. KISS and YAGNI principles apply here.

NikolaySlavov-96 replied:
Thank you very much for your attentiveness.

vgechev commented on Services/Repositories/PushNotification/Repository/PushNotificationRepository.cs:
```diff
@@ -0,0 +1,45 @@
+
+            await SaveChangesAsync();
+
+            return notificationData;
```
Same here, refer to my previous comment.

vgechev commented on Services/Repositories/PushNotification/Repository/PushNotificationRepository.cs:
```diff
@@ -0,0 +1,45 @@
+        public Task<bool> CheckDeviceTokenExists(PushNotificationDTO notificationDTO, int userId) =>
+            dbContext.PushNotifications.AnyAsync(p => p.Token == notificationDTO.Token && p.UserId == userId);
+
+        public Task<PushNotificationEntity?> GetPushNotificationIdByDeviceId(string deviceId, int userId)
```
Rename to IsNotificationEnabledForDeviceId( ... ) and make a .Select to return only what you need - a boolean.

vgechev commented on Services/Repositories/PushNotification/Repository/PushNotificationRepository.cs:
```diff
@@ -0,0 +1,45 @@
+            return dbContext.PushNotifications.FirstOrDefaultAsync(r => r.UserDevice.DeviceId == deviceId && r.UserId == userId);
+        }
+
+        public Task SaveChangesAsync() =>
```
you have savechanges, but are also calling saveChanges separately in the methods..? Why?

NikolaySlavov-96 replied:
Apparently, I am quite unfamiliar. We will discuss it when we talk.

vgechev commented on Services/Repositories/OperationSystem/Repository/OperationSystemRepository.cs:
```diff
@@ -0,0 +1,23 @@
+    {
+        public async Task<int?> GetOperationSystemId(string operationSystem)
+        {
+            OperationSystemEntity result = await dbContext.OperationSystems.FirstOrDefaultAsync(o => o.TypeOS == operationSystem);
```
There is a warning here. Add ? to OperationSystemEntity, like so `OperationSystemEntity?`

vgechev commented on Services/Repositories/OperationSystem/Repository/OperationSystemRepository.cs:
```diff
@@ -0,0 +1,23 @@
+        public async Task<int?> GetOperationSystemId(string operationSystem)
+        {
+            OperationSystemEntity result = await dbContext.OperationSystems.FirstOrDefaultAsync(o => o.TypeOS == operationSystem);
+            if (result != null)
```
you can remove the IF and shorten the whole part of the code to just this:

OperationSystemEntity? entity = ....;
return result?.Id;

vgechev replied:
And again, I have no idea why you wouldn't just use an ENUM instead. It would simplify the code by a lot.

vgechev commented on Turbo/Controllers/PushNotification/Controller/PushNotificationController.cs:
```diff
@@ -0,0 +1,50 @@
+        public async Task<IActionResult> GetPushNotificationDeviceData()
+        {
+            var result = await pushNotificationService.GetPushNotificationData(User.DeviceId(), User.Id());
+            if (result.IsSuccess)
```
Just a suggestion - I'm pretty sure we can make a middleware to handle this and not have to duplicate these "if (isSuccess) ... else ..."

NikolaySlavov-96 replied:
I agree with you, but i am following the other controllers. If it is devised to make the change, we will create a PR specifically for this change, and it will be applied.

vgechev commented on Turbo/Program.cs:
```diff
@@ -61,7 +62,12 @@ public static void Main(string[] args)
+			var seederRunner = service.GetRequiredService<SeederRunner>();
+
+			context.ApplyMigrations();
+			seederRunner.RunSeeders(context);
```
no :D

vgechev reviewed (changes requested):
1. The seeding code should not be part of the normal app execution as this can cause concurrency issues when multiple instances are running and would also require the app having permission to modify the database schema, if needed. Since the seed is very small, you could've just made it in the OnModelCreating method, example:
![image](https://github.com/user-attachments/assets/00cfd407-bfd6-457d-800f-179485a69e89)


Or if you want to make it more abstract, you could create a class with the seed data, an extenstion method (e.g. called Seed) to ModelBuilder (the param from OnModelCreating) and just call modelBuilder.Seed() in the OnModelCreating method. This way it will be very clean, I can provide examples if you want, but I'm sure a quick google search will show you the same result.
The solution with these seeders is more confusing and also you are calling "SaveChanges" which should be a red flag because in seeding you should not be calling it. P.S. you should usually use the Async method for SaveChanges =).

There is a third option, you can do it in the migration directly, like so:
![image](https://github.com/user-attachments/assets/3937c25c-cc17-4e7a-9d0f-92d92e39fc9f)



2. Honestly, I'm kind of confused why you would even need the OperationSystem table since it will probably only have 2, at most 3, records and you could've easily just used an enum instead. I don't see any benefit in creating a table.

3. The naming of some of the classes and params can be improved. I haven't commented all of them, please review.

4. This is a question for everyone - shouldn't the repositories be in the Database project?

NikolaySlavov-96 commented:
@vgechev Thank you very much for the comments. I learned a lot of useful things. Those with which i agree will be changed very soon. For the others on which there is a comment, i expect a resolution before doing anything.

NikolaySlavov-96 committed:
chore: Remove seeders from the project

NikolaySlavov-96 committed:
refactor: Remove OperatingSystem table and update pushNotification table to use enum

NikolaySlavov-96 committed:
refactor: Resolve code comments, rename elements, and remove unused data

NikolaySlavov-96 committed:
refactor: Rename elements, add summaries, and update PushNotification DB logic on add and save methods

NikolaySlavov-96 committed:
refactor: Rename method and variables for clarity and consistency

- Renamed the method DoesDeviceTokenExist to IsTokenExistsForUser to better reflect its purpose and improve readability.

NikolaySlavov-96 committed:
refactor: rename PushNotification to DeviceNotificationConfig across services, controllers, and repositories

NaumVitanov commented on Services/Repositories/DeviceNotificationConfig/Repository/DeviceNotificationConfigRepository.cs:
```diff
@@ -0,0 +1,42 @@
+            return true;
+        }
+
+        public Task<bool> IsTokenExistsForUser(string deviceToken, int userId) =>
```
This way it's grammatically incorrect. DoesTokenExistForUser, would be correct, even though it does not follow the "Is,Has" pattern for booleans

NaumVitanov commented on Services/DeviceNotificationConfigService/Service/DeviceNotificationConfigService.cs:
```diff
@@ -0,0 +1,61 @@
+                return Result.Failure("Something went wrong with your device token");
+            };
+
+            int userDeviceId = await userRepo.GetDeviceByDeviceId(deviceId);
```
The naming of the method is wrong. Please change it to show that it's getting the userDeviceId

NikolaySlavov-96 committed:
comm: Resole PR comment (naming)

NaumVitanov reviewed (approved):
Thanks for complying with our requests! Good job!
